nomad/structs/structs.go

```diff
@@ -5010,6 +5116,49 @@ func (a *Allocation) TerminalStatus() bool {
 	}
 }
 
+// ShouldReschedule returns if the allocation is eligible to be rescheduled according
+// to its status and ReschedulePolicy
+func (a *Allocation) ShouldReschedule(reschedulePolicy *ReschedulePolicy) bool {
```

Comment:
You may want to consider allowing these two functions to take the failure time of the allocation instead of using time.Now(). This has two benefits, it is easier to test and in the future we may have a better means of detecting the time the allocation failed on the client so the scheduler can use that time rather than the time that the scheduler calls this function.